// vue-simple-select/option.js
module.exports = {
  mixins: [require("vue-mixins/class")],
  props: {
    "class": {
      "default": function() {
        return [];
      }
    },
    value: null,
    disabled: {
      type: Boolean,
      "default": false
    }
  },
  computed: {
    mergeClass: function() {
      return {
        active: this.active,
        selected: this.selected,
        disabled: this.disabled
      };
    },
    isMultiple: function() {
      return this.$parent.multiple;
    }
  },
  data: function() {
    return {
      active: false,
      selected: false,
      isOption: true
    };
  },
  methods: {
    onClick: function(e) {
      if (!e.defaultPrevented) {
        return this.$parent.onClick(e, this);
      }
    },
    onMouseDown: function(e) {
      if (!(this.disabled || e.defaultPrevented)) {
        return this.$parent.onMouseDown(e, this);
      }
      return false;
    }
  }
};

if (module.exports.__esModule) module.exports = module.exports.default
;(typeof module.exports === "function"? module.exports.options: module.exports).template = "<li :class=computedClass @mousedown=onMouseDown @click=onClick><slot name=img></slot><span><template v-if=isMultiple><input type=checkbox read-only=read-only :checked=active><label></label></template><slot>{{value}}</slot></span>"
