// vue-materialize/card.js
(function() {
  var Velocity, c, cancel, card;

  card = require("vue-card");

  Velocity = require("velocity-animate");

  cancel = function(el) {
    return Velocity(el, "stop");
  };

  c = {
    template: card.template,
    mixins: card.mixins,
    props: card.props,
    computed: card.computed,
    data: card.data,
    methods: card.methods,
    transitions: {
      card: {
        css: false,
        leaveCancelled: cancel,
        enterCancelled: cancel,
        beforeEnter: function() {},
        enter: function(el, done) {
          Velocity.hook(el, "translateY", "100%");
          return Velocity(el, {
            translateY: 0
          }, {
            duration: 300,
            easing: "easeInOutQuad",
            queue: false,
            complete: done
          });
        },
        leave: function(el, done) {
          return Velocity(el, {
            translateY: "100%"
          }, {
            duration: 300,
            easing: "easeInOutQuad",
            queue: false,
            complete: done
          });
        }
      }
    }
  };

  module.exports = c;

}).call(this);
