// vue-materialize/collapsible-item.js
(function() {
  var Velocity, cancel, ci, comp;

  Velocity = require("velocity-animate");

  cancel = function(el) {
    return Velocity(el, "stop");
  };

  ci = require("vue-collapsible/collapsible-item");

  comp = {
    template: ci.template,
    mixins: ci.mixins,
    props: ci.props,
    computed: ci.computed,
    data: ci.data,
    methods: ci.methods,
    transitions: {
      collapsible: {
        css: false,
        leaveCancelled: cancel,
        enterCancelled: cancel,
        enter: function(el, done) {
          return Velocity(el, "slideDown", {
            duration: 350,
            easing: "easeOutQuart",
            queue: false,
            complete: done
          });
        },
        leave: function(el, done) {
          return Velocity(el, "slideUp", {
            duration: 350,
            easing: "easeOutQuart",
            queue: false,
            complete: done
          });
        }
      }
    },
    ready: function() {
      if (this.isOpened) {
        return this.$el.children[1].style.display = "block";
      }
    }
  };

  module.exports = comp;

}).call(this);
