// vue-materialize/dropdown.js
(function() {
  var Velocity, cancel, dropdown;

  Velocity = require("velocity-animate");

  dropdown = require("vue-comps-dropdown");

  cancel = function(el) {
    return Velocity(el, "stop");
  };

  module.exports = {
    template: dropdown.template,
    mixins: dropdown.mixins,
    props: dropdown.props,
    data: dropdown.data,
    computed: {
      cAnchor: dropdown.computed.cAnchor,
      mergeStyle: dropdown.computed.mergeStyle,
      mergeClass: function() {
        return ["dropdown-content"];
      }
    },
    methods: dropdown.methods,
    ready: dropdown.ready,
    beforeDestroy: dropdown.beforeDestroy,
    transitions: {
      dropdown: {
        css: false,
        enter: function(el, done) {
          Velocity(el, "slideDown", {
            duration: 300,
            queue: false,
            easing: 'easeOutCubic',
            complete: done
          });
          return Velocity(el, {
            opacity: 1
          }, {
            duration: 300,
            queue: false,
            easing: 'easeOutSine'
          });
        },
        leave: function(el, done) {
          return Velocity(el, {
            opacity: 0
          }, {
            duration: 225,
            queue: false,
            easing: 'easeOutSine',
            complete: done
          });
        },
        leaveCancelled: cancel,
        enterCancelled: cancel
      }
    }
  };

}).call(this);
