// vue-materialize/material-box.js
(function() {
  var Velocity, b, box, cancel;

  Velocity = require("velocity-animate");

  require("./overlay");

  cancel = function(el) {
    return Velocity(el, "stop");
  };

  box = require("vue-zoombox");

  b = {
    template: box.template,
    mixins: box.mixins,
    props: box.props,
    computed: box.computed,
    data: box.data,
    watch: box.watch,
    methods: box.methods,
    beforeCompile: box.beforeCompile,
    ready: box.ready,
    beforeDestroy: box.beforeDestroy,
    transitions: {
      zoombox: {
        css: false,
        leaveCancelled: cancel,
        enterCancelled: cancel,
        enter: function(el, done) {
          Velocity.hook(el, "scale", this.imgScale);
          return this.$nextTick((function(_this) {
            return function() {
              return Velocity(el, {
                scale: _this.imgScale * _this.scale,
                top: _this.relPos.top,
                left: _this.relPos.left
              }, {
                duration: 275,
                queue: false,
                ease: "easeOutQuad",
                complete: done
              });
            };
          })(this));
        },
        leave: function(el, done) {
          Velocity.hook(el, "scale", this.imgScale * this.scale);
          return Velocity(el, {
            scale: this.imgScale,
            top: 0,
            left: 0
          }, {
            duration: 200,
            queue: false,
            ease: "easeOutQuad",
            complete: done
          });
        }
      },
      zoomboxCaption: {
        css: false,
        leaveCancelled: cancel,
        enterCancelled: cancel,
        enter: function(el, done) {
          return Velocity(el, {
            opacity: 1
          }, {
            duration: 275,
            queue: false,
            ease: "easeOutQuad",
            complete: done
          });
        },
        leave: function(el, done) {
          return Velocity(el, {
            opacity: 0
          }, {
            duration: 200,
            queue: false,
            ease: "easeOutQuad",
            complete: done
          });
        }
      }
    }
  };

  module.exports = b;

}).call(this);
