// vue-materialize/toast.js
var Velocity;

Velocity = require("velocity-animate");

module.exports = {
  components: {
    "drag-handle": require("vue-drag-handle")
  },
  props: {
    options: {
      type: Object
    }
  },
  data: function() {
    return {
      style: {
        top: "35px",
        opacity: 0,
        position: "relative",
        left: 0
      }
    };
  },
  methods: {
    move: function(dX) {
      this.style.left = dX + 'px';
      return this.style.opacity = 1 - Math.abs(dX / 125);
    },
    close: function() {
      return this.options.close();
    },
    abort: function() {
      Velocity(this.$el, "stop");
      return Velocity(this.$el, {
        "left": 0,
        opacity: 1
      }, {
        duration: 300,
        easing: 'easeOutCubic'
      });
    }
  }
};

if (module.exports.__esModule) module.exports = module.exports.default
;(typeof module.exports === "function"? module.exports.options: module.exports).template = "<div :style=style>{{options.text}}<drag-handle @move=move @max=close @aborted=abort :max-right=125 :max-left=125 style=width:100%;left:0></drag-handle></div>"
