// vue-toggle/toggle.js
module.exports = {
  mixins: [require("vue-mixins/class")],
  props: {
    "class": {
      "default": function() {
        return [];
      }
    },
    onClass: {
      type: String,
      "default": "on"
    },
    offClass: {
      type: String,
      "default": "off"
    },
    isOn: {
      type: Boolean,
      "default": false
    },
    disabled: {
      type: Boolean,
      "default": false
    }
  },
  computed: {
    mergeClass: function() {
      if (this.isOn) {
        return [this.onClass];
      } else {
        return [this.offClass];
      }
    }
  },
  methods: {
    toggle: function(e) {
      if ((e != null) && e.defaultPrevented) {
        return;
      }
      if (e != null) {
        if (typeof e.preventDefault === "function") {
          e.preventDefault();
        }
      }
      if (this.isOn) {
        this.$emit("off", e);
      } else {
        this.$emit("on", e);
      }
      this.$emit("toggle", e);
      return this.isOn = !this.isOn;
    }
  }
};

if (module.exports.__esModule) module.exports = module.exports.default
;(typeof module.exports === "function"? module.exports.options: module.exports).template = "<button @click=toggle v-bind:class=computedClass v-bind:disabled=disabled><slot>Toggle:</slot><slot name=on v-if=isOn>on</slot><slot name=off v-else=v-else>off</slot></button>"
