// vue-mixins/transition2.js
(function() {
  module.exports = {
    mixins: [require("./vue")],
    computed: {
      cTransition: function() {
        return this.processTransition(this.transition);
      },
      cTransitionGroup: function() {
        return this.processTransitionGroup(this.transition);
      }
    },
    methods: {
      processTransitionGroup: function(name, options) {
        if (options == null) {
          options = {};
        }
        if (options.defaultName == null) {
          options.defaultName = "transition-group";
        }
        return this.processTransition(name, options);
      },
      processTransition: function(name, options) {
        var comp, defaultName, parent, ref, ref1;
        if (options == null) {
          options = {};
        }
        parent = (ref = options.parent) != null ? ref : this.$parent, defaultName = (ref1 = options.defaultName) != null ? ref1 : "transition";
        if (parent == null) {
          parent = this;
        }
        comp = this.Vue.util.resolveAsset(parent.$options, 'components', name);
        if (parent !== this && (comp == null)) {
          comp = this.Vue.util.resolveAsset(this.$options, 'components', name);
        }
        if (comp != null) {
          this.$options.components[name] = comp;
        } else {
          name = defaultName;
        }
        return name;
      }
    }
  };

}).call(this);
