// vue-side-nav/side-nav.js
module.exports = {
  components: {
    "drag-handle": require("vue-drag-handle")
  },
  created: function() {
    return this.overlay = require("vue-overlay")(this.Vue);
  },
  mixins: [require("vue-mixins/vue"), require("vue-mixins/onWindowResize"), require("vue-mixins/setCss"), require("vue-mixins/isOpened"), require("vue-mixins/class"), require("vue-mixins/style")],
  props: {
    "id": {
      type: String
    },
    "class": {
      "default": function() {
        return ["side-nav"];
      }
    },
    "style": {
      "default": function() {
        return [];
      }
    },
    "width": {
      type: Number,
      coerce: Number,
      "default": 240
    },
    "opacity": {
      type: Number,
      "default": 0.5,
      coerce: Number
    },
    "right": {
      type: Boolean,
      "default": false
    },
    "notDismissable": {
      type: Boolean,
      "default": false
    },
    "closeOnClick": {
      type: Boolean,
      "default": false
    },
    "fixed": {
      type: Boolean,
      "default": false
    },
    "fixedScreenSize": {
      type: Number,
      coerce: Number,
      "default": 992
    },
    "transition": {
      type: Function,
      "default": function(arg) {
        var cb, el, style;
        el = arg.el, style = arg.style, cb = arg.cb;
        this.position = style[this.side] + "px";
        return cb();
      }
    },
    "isFixed": {
      type: Boolean,
      "default": false
    },
    "zIndex": {
      type: Number,
      "default": 1000,
      coerce: Number
    }
  },
  computed: {
    side: function() {
      if (this.right) {
        return "right";
      } else {
        return "left";
      }
    },
    mergeClass: function() {
      return {
        fixed: this.fixed
      };
    },
    mergeStyle: function() {
      var style;
      style = {
        position: "fixed",
        width: this.width + "px",
        top: 0,
        margin: 0,
        height: "100%",
        zIndex: this.overlayZIndex + 1,
        boxSizing: "border-box",
        transform: "translateX(0)"
      };
      if (this.position) {
        style[this.side] = this.position;
      }
      return style;
    }
  },
  watch: {
    fixed: "processFixed",
    fixedScreenSize: "processFixed"
  },
  data: function() {
    return {
      position: -1 * (this.width + 10) + "px",
      overlayZIndex: 1001
    };
  },
  methods: {
    makeFixed: function(fixed) {
      if (fixed !== this.isFixed) {
        this.isFixed = fixed;
        this.setParentMargin(fixed);
        return this.$emit("fixed", fixed);
      }
    },
    setParentMargin: function(fixed, side) {
      var el, i, len, ref, results, width;
      if (fixed == null) {
        fixed = this.isFixed;
      }
      if (side == null) {
        side = this.side;
      }
      if (this.$el.parentElement) {
        if (fixed) {
          if (this.computedStyle[1] != null) {
            width = this.computedStyle[1].width;
          }
          if (width == null) {
            width = this.computedStyle[0].width;
          }
        } else {
          width = null;
        }
        ref = this.$el.parentElement.children;
        results = [];
        for (i = 0, len = ref.length; i < len; i++) {
          el = ref[i];
          if (el !== this.$els.nav) {
            results.push(this.setCss(el, "margin-" + side, width));
          } else {
            results.push(void 0);
          }
        }
        return results;
      }
    },
    processFixed: function() {
      var isFixed;
      if (this.fixed) {
        isFixed = window.innerWidth > this.fixedScreenSize;
        this.makeFixed(isFixed);
        if (isFixed) {
          this.position = 0;
        } else {
          this.position = -1 * (this.width + 10) + "px";
        }
        this.disposeWindowResize = this.onWindowResize((function(_this) {
          return function() {
            if (window.innerWidth > _this.fixedScreenSize) {
              if (!_this.isFixed) {
                if (_this.isOpened) {
                  _this.close(true);
                  _this.wasOpened = true;
                } else {
                  _this.show(false);
                }
                return _this.makeFixed(true);
              }
            } else {
              if (_this.isFixed) {
                if (_this.wasOpened) {
                  _this.open(true);
                } else {
                  _this.hide(false);
                }
                return _this.makeFixed(false);
              }
            }
          };
        })(this));
      } else {
        this.isFixed = false;
        if (typeof this.disposeWindowResize === "function") {
          this.disposeWindowResize();
        }
      }
      return this.setParentMargin();
    },
    onClick: function(e) {
      if (this.closeOnClick) {
        return this.dismiss(e);
      }
    },
    dismiss: function(e) {
      if (!e.defaultPrevented) {
        if (!this.notDismissable && !this.isFixed) {
          this.close();
          return e.preventDefault();
        }
      }
    },
    move: function(position) {
      var fac;
      fac = this.right ? -1 : 1;
      return this.position = -this.width + fac * position + "px";
    },
    show: function(animate) {
      var style;
      if (animate == null) {
        animate = true;
      }
      this.$emit("before-enter");
      if (animate) {
        style = {};
        style[this.side] = 0;
        return this.transition({
          el: this.$els.nav,
          style: style,
          cb: (function(_this) {
            return function() {
              _this.setCss(_this.$els.nav, "transform", "translateX(0)");
              _this.setOpened();
              return _this.$emit("after-enter");
            };
          })(this)
        });
      } else {
        this.position = 0;
        this.setOpened();
        return this.$emit("after-enter");
      }
    },
    hide: function(animate) {
      var style;
      if (animate == null) {
        animate = true;
      }
      this.wasOpened = false;
      this.$emit("before-leave");
      if (animate) {
        style = {};
        style[this.side] = -1 * (this.width + 10);
        return this.transition({
          el: this.$els.nav,
          style: style,
          cb: (function(_this) {
            return function() {
              _this.setClosed();
              return _this.$emit("after-leave");
            };
          })(this)
        });
      } else {
        this.position = -1 * (this.width + 10) + "px";
        this.setClosed();
        return this.$emit("after-leave");
      }
    },
    open: function(restoreOverlay) {
      var close, ref, zIndex;
      if (this.opened && !restoreOverlay) {
        return;
      }
      ref = this.overlay.open({
        zIndex: this.zIndex,
        opacity: this.opacity,
        onBeforeClose: (function(_this) {
          return function() {
            return _this.close();
          };
        })(this)
      }), zIndex = ref.zIndex, close = ref.close;
      this.overlayZIndex = zIndex;
      this.closeOverlay = close;
      if (!restoreOverlay) {
        this.show();
        return this.$emit("toggled", true);
      }
    },
    close: function(restoreNav) {
      if (!this.opened) {
        return;
      }
      if (typeof this.closeOverlay === "function") {
        this.closeOverlay(false);
      }
      this.closeOverlay = null;
      if (!restoreNav) {
        this.hide();
        return this.$emit("toggled", false);
      }
    },
    toggle: function() {
      if (this.isFixed) {
        return this.isOpened = this.opened;
      } else {
        if (this.opened) {
          return this.close();
        } else {
          return this.open();
        }
      }
    }
  },
  ready: function() {
    return this.processFixed();
  }
};

if (module.exports.__esModule) module.exports = module.exports.default
;(typeof module.exports === "function"? module.exports.options: module.exports).template = "<drag-handle @move=move @max=open(false) @aborted=hide v-bind:disabled=\"isOpened || isFixed\" v-bind:max-right=\"right ? null : width\" v-bind:max-left=\"right ? width : null\" v-bind:z-index=overlayZIndex v-bind:style=\"{width: '20px',left:right ? null : 0,right:right ? 0 : null}\"></drag-handle><drag-handle @move=move @max=close(false) @aborted=show v-bind:disabled=\"!isOpened || isFixed\" v-bind:max-right=\"right ? width : null\" v-bind:max-left=\"right ? null : width\" v-bind:offset=\"right ? -width : width\" v-bind:z-index=overlayZIndex v-bind:style={left:0,right:0} @clean-click=dismiss></drag-handle><ul v-bind:id=id v-el:nav=v-el:nav @click=onClick @keyup.esc=dismiss v-bind:style=computedStyle v-bind:class=computedClass><slot>No content</slot></ul>"
